src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/UnsafeAccessorAttribute.cs

```diff
@@ -0,0 +1,89 @@
+    /// <remarks>
+    /// This attribute should be applied on a <code>extern static</code> method.
+    /// The implementation of the <code>extern static</code> method annotated with
+    /// this attribute will be provided by the runtime based on the information in
+    /// the attribute and the signature of the method that the attribute is applied to.
+    /// The runtime will try to find the matching method or field and forward the call
+    /// to it. If the matching method or field is not found, the body of the <code>extern</code>
+    /// method will throw <see cref="MissingFieldException" /> or <see cref="MissingMethodException" />.
+    ///
+    /// For <see cref="UnsafeAccessorKind.Method"/>/<see cref="UnsafeAccessorKind.StaticMethod"/> and
+    /// <see cref="UnsafeAccessorKind.Field"/>/<see cref="UnsafeAccessorKind.StaticField"/>, the type of
+    /// the first argument of the annotated <code>extern</code> method identifies the owning type.
+    /// The value of the first argument is treated as <code>this</code> pointer for instance fields and methods.
+    /// The first argument must be passed as <code>ref</code> for instance fields and methods on structs.
+    /// The value of the first argument is not used by the implementation for <code>static</code> fields and methods.
+    ///
+    /// The generic parameters of the <code>extern static</code> method are concatenation of the type and
+    /// method generic arguments of the target method. For example,
+    /// <code>extern static void Method1&lt;T1, T2&gt;(Class1&lt;T1&gt; @this)</code>
+    /// can be used to call <code>Class1&lt;T1&gt;.Method1&lt;T2&gt;()</code>. The generic constraints of the
+    /// <code>extern static</code> method must match generic constraints of the target type, field or method.
+    ///
+    /// Return type is considered for the signature match. modreqs and modopts are not considered for the signature match.
+    /// </remarks>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+    public sealed class UnsafeAccessorAttribute : Attribute
+    {
+        /// <summary>
+        /// Instantiates an <see cref="UnsafeAccessorAttribute"/> providing access to an API of kind <see cref="UnsafeAccessorKind"/>.
+        /// </summary>
+        /// <param name="kind">The kind of the target to provide access.</param>
+        public UnsafeAccessorAttribute(UnsafeAccessorKind kind)
+            => Kind = kind;
+
+        /// <summary>
+        /// Kind of API to provide access.
+        /// </summary>
+        public UnsafeAccessorKind Kind { get; }
+
+        /// <summary>
+        /// Name of the API to provide access.
+        /// </summary>
+        /// <remarks>
+        /// The name defaults to the annotated method name if not specified.
+        /// The name must be unset/<code>null</code> for <see cref="UnsafeAccessorKind.Constructor"/>.
+        /// </remarks>
+        public string? Name { get; set; }
```

Comment:
Why wouldn't it just be ignored if the Kind is Constructor?

Reply:
In this case if we require it to be null we can relax at some point in the future, at least that is why I am forcing the situation.